### DISCUSSION TITLE
How can I set pointer position on SwapChainPanel in C#?

### DISCUSSION THREAD

**Discussion by vmx17** [microsoft/WindowsAppSDK#3427] [Q&A] [answered]
Now I making prototype of drawing program using WinUI 3 desktop in C#.
Just, how can I set pointer position on SwapChainPanel in C#? means, move mouse cursor by code.
Though I think this kind of question is very basic on WindowsForms, WPF or UWP but could not find answer on WinUI3.
I tried some codes like;
`(Application.Current as App).CurrentWindow.CoreWindow.PointerPosition = new(x, y);`
but I have no luck. In this case, where `(Application.Current as App).CurrentWindow` is set manually (because it is not static), but `CoreWindow` is null. So this end with NullReference Exception. I have a local position on SwapChainPanel and want move mouse cursor there.

Possibly, I'm expecting code as property something like;
`public some_Point_type CursorPosition {get=>m_position; set=>SetProperty(ref m_position, value)};`
in ViewModel.

**Comment by DarranRowe** [comment]
There is nothing overly special about doing this here since WinUI 3 is a library running in a desktop application, so any desktop application method for doing this will work. Of course, because this is a desktop environment, UWP features like CoreWindow and the UWP Dispatcher are totally unavailable.
Anyway, you should be looking at System.Windows.Forms.Cursor or the Windows API SetCursorPos function.

**Reply by vmx17** [reply]
Thank you. So, that mean even in WinUI3 desktop, we should refer `WindowsForms` or use `user32.dll` (via `DllImport`). I couldn't think of it.

**Comment by castorix** [comment, marked as answer]
You can use APIs like _SetCursorPos_, _SendInput_ or _mouse_event_ and convert Screen to Client coordinates

Something like (test to position at 10, 10 inside a SwapChainPanel named _scp1_:

                IntPtr hWndMain = WinRT.Interop.WindowNative.GetWindowHandle(this);
                Windows.Graphics.PointInt32 pt = new Windows.Graphics.PointInt32(10, 10);
                ClientToScreen(hWndMain, ref pt);
                var gt = scp1.TransformToVisual((UIElement)this.Content);
                var ptNew = gt.TransformPoint(new Windows.Foundation.Point(0, 0));
                pt.X += (int)ptNew.X;
                pt.Y += (int)ptNew.Y;
                MoveCursor(pt.X, pt.Y);

with :

        [DllImport("User32.dll", ExactSpelling = true, EntryPoint = "SetCursorPos", CharSet = CharSet.Unicode)]
        public static extern bool MoveCursor(int X, int Y);

        [DllImport("User32.dll", SetLastError = true, CharSet = CharSet.Unicode)]
        public static extern bool ClientToScreen(IntPtr hWnd, ref Windows.Graphics.PointInt32 lpPoint);

**Reply by vmx17** [reply]
Thank you. After all, I've got `IntPtr` of `hWnd` for every window (bacause WindowsAppSdk can handle multiple windows), then it should be used as traditional way... I was prejudiced that WinUI3 has almost same way as UWP.
